Tests/BuildTests/IncrementalBuildTests.swift

```diff
@@ -54,22 +181,293 @@ final class IncrementalBuildTests: XCTestCase {
+
+            var logs = BuildLog()
+            logs[.initial] = BuildResult(try? executeSwiftBuild(prefix, printIfError: true), path: prefix)
+            // remove the dependency
+            try localFileSystem.writeFileContents(packagePath, bytes: packageNoDep.bytes)
+            logs[.withoutDependency] = BuildResult(try? executeSwiftBuild(prefix, printIfError: true), path: prefix)
+            // put it back
+            try localFileSystem.writeFileContents(packagePath, bytes: packageDep.bytes)
+            logs[.withDependency] = BuildResult(try? executeSwiftBuild(prefix, printIfError: true), path: prefix)
+            // Once more, to verify it doesn't do anything for no changes
+            logs[.null] = alwaysDoRedundancyCheck ? BuildResult(try? executeSwiftBuild(prefix, printIfError: true), path: prefix) : nil
+            
+            // Check for unexpected build failures
+            checkForNils(logs).forEach {
+                XCTFail("\(line()): \(IncBuildErrorMessages.unexpectedBuildFailure): \($0)")
+            }
+            
+            // Check the logs
+            XCTAssert(logs[.initial]!.isNotNullBuild,                                       "\(line()): \(IncBuildErrorMessages.unexpectedNullBuild)")
+            // 1st rebuild... there's nothing actually using the dependency, so
+            // this should be a null build
+            XCTAssert(logs[.withoutDependency]!.isNullBuild,                                "\(line()): \(IncBuildErrorMessages.unexpectedBuild)")
+            // 2nd rebuild... should still be a null build
+            XCTAssert(logs[.withDependency]!.isNullBuild,                                   "\(line()): \(IncBuildErrorMessages.unexpectedBuild)")
+            XCTAssert(logs[.initial]!.buildProducts == logs[.withDependency]!.buildProducts,"\(line()): \(IncBuildErrorMessages.builtProductsDiffer)")
+            
+            // There shouldn't be anything for build 4 to do
+            XCTAssert(logs[.null]?.isNullBuild ?? true,                                     "\(line()): \(IncBuildErrorMessages.unexpectedBuild)")
+        }
+    }
+    
+    func testAddRemoveFiles() {
+        fixture(name: "IncrementalBuildTests/FileAddRemoveTest") { prefix in
+            let extraFilePath = prefix.appending(components: "Sources", "FileTesterExt.swift")
+            let extraFile = BufferedOutputByteStream()
+            
+            // We want to be careful with the filesystem here. There's not much
+            // we can do to recover from an error writing the file back out, but
+            // if we get an error reading the file in in the first place,
+            // something has gone really wrong.
+            do {
+                extraFile <<< (try localFileSystem.readFileContents(extraFilePath))
+            } catch let error {
+                let message = "\(line())caught \"\(error)\" trying to read \(extraFilePath.asString). Aborting test"
```

Comment:
Any reason this test is careful but not the others? One trick that I use is to just mark the test case itself as "throws" and let the error throw all the way out of the test, if isn't the kind of error I am actually writing the test to normally catch with a richer diagnostic.